### PR TITLE
Improve performance, reliability of firestore backend

### PR DESCRIPTION
**Pull request by xacrimon**
This PR makes three improvements to the firestore backend for events. The first one is batched requests with type filtering which reduces round trips from the frontend when viewing the audit log with event type queries. The second improvement is defining a previously not set variable Firestore which is currently defaulted but the API does not specify behaviour. The third improvement is to use the wait API from Google to avoid polling on index creation when possible.

This PR has been extensively manually tested and existing index and query tests pass.

### CODE REVIEW

**Review by fspmarshall — APPROVED** [review]
Nit: it looks like previously a periodic status was logged, but with these changes logs are only generated at the start of each "wait".  Might be worth adding some kind of periodic log message back in, just so that in case an operation takes a while it is clear from the logs what teleport is waiting on.

File: lib/backend/firestore/firestorebk.go [hunk 1]
```diff
@@ -768,36 +768,30 @@ func EnsureIndexes(ctx context.Context, adminSvc *apiv1.FirestoreAdminClient, tu
...
-	// finish. Also, there should ideally be a timeout on index creation.
-
-	// check for statuses and block
-	for len(tuplesToIndexNames) != 0 {
-		select {
-		case <-time.After(timeInBetweenIndexCreationStatusChecks):
-		case <-ctx.Done():
-			return trace.ConnectionProblem(ctx.Err(), "context timed out or canceled")
+	for i, operation := range tasks {
+		err := waitOnIndexCreation(ctx, l, operation, tuples[i])
```
Comment (nklaassen) [inline lib/backend/firestore/firestorebk.go:776]: looks like this counts on `tasks` and `tuples` lining up by index/having the same length, but above you can skip adding the operation to `tasks` if it is nil

Comment (xacrimon) [reply lib/backend/firestore/firestorebk.go]: Fixed by merging operations and tuples into one struct

**Review by nklaassen — APPROVED** [review]

**Review by russjones — APPROVED** [review]
Bot.

**Review by klizhentas — APPROVED** [review]

**Pull request merged**